### ISSUE TITLE
Polyfill breaks inline script load order

### ISSUE THREAD

**Issue by elliotcm** [guybedford/es-module-shims#308]
Hello.

I've read through the readme and similar github issues to this but nothing quite matches, please let me know if this is actually a duplicate.

I'm converting from webpacker to importmaps and having trouble with some timing issues.  Given this (mostly generated by Rails) `head`:
```
<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-0evHe/X+R7YkIZDRvuzKMRqM+OrBnVFBL6DOitfPri4tjfHxaWutUpFmBp4vmVor" crossorigin="anonymous">
    <link rel="stylesheet" href="/assets/munificent-ui-a72913f4c604bc8f97dd576fc8777bf029401f6af0e14b6a6c9d3874acfe73e4.css" media="all" />
    <script type="importmap" data-turbo-track="reload">{
  "imports": {
    "munificent_ui/application": "/assets/munificent_ui/application-3f3cbee6448455fc9bc308b3b87f7ef84ad40996e27701deef188e8c01a31742.js",
    "@rails/ujs": "https://ga.jspm.io/npm:@rails/ujs@7.0.3/lib/assets/compiled/rails-ujs.js",
    "@rails/activestorage": "https://ga.jspm.io/npm:@rails/activestorage@7.0.3/app/assets/javascripts/activestorage.esm.js",
    "jquery-ujs": "https://ga.jspm.io/npm:jquery-ujs@1.2.3/src/rails.js",
    "munificent_accounts/application": "/assets/munificent_accounts/application-120c743b653969372053b84133aac8b96c4259024616be425d478969cf73e109.js",
    "munificent_payments/application": "/assets/munificent_payments/application-120c743b653969372053b84133aac8b96c4259024616be425d478969cf73e109.js"
  }
}</script>
<link rel="modulepreload" href="/assets/munificent_ui/application-3f3cbee6448455fc9bc308b3b87f7ef84ad40996e27701deef188e8c01a31742.js">
<link rel="modulepreload" href="/assets/munificent_accounts/application-120c743b653969372053b84133aac8b96c4259024616be425d478969cf73e109.js">
<link rel="modulepreload" href="/assets/munificent_payments/application-120c743b653969372053b84133aac8b96c4259024616be425d478969cf73e109.js">
<script src="/assets/es-module-shims.min-606ae9c3279013fe751cee30f719a592f759e705edb66496812f3d9dbce3d850.js" async="async" data-turbo-track="reload"></script>
<script type="module">import "munificent_ui/application"</script>

    <script type="text/javascript">
      function onReady(callback) {
        if (document.readyState != 'loading'){
          callback();
        } else {
          document.addEventListener('DOMContentLoaded', callback);
        }
      }
    </script>

      <script src="https://js.stripe.com/v3/"></script>
  <script src="https://www.paypal.com/sdk/js?client-id=AakBwrFPLdPU4-oKX939jxpGlL-0_J9E10z1dSapKjp2G-o8pte9OjKw8bBlQEHKbyFWU9WqAdTMgupz&currency=GBP&integration-date=2022-04-08&disable-funding=card"></script>
  <script type="text/javascript">
    onReady(function() {
      window.payments.addStripe(document.getElementById('stripe-donate-button'));
      window.payments.addPaypal(document.getElementById('paypal-donate-button'));
      // Packs.application.enableCharitySplit(document.getElementsByClassName('charity-split')[0]);
    });
  </script>
```
in a non-polyfilled browser the tag `<script type="module">import "munificent_ui/application"</script>` would be ([by default](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-defer)) deferred until just before `DOMContentLoaded`, meaning that the inline script would execute after the module had loaded (which is what sets `window.payments`).

As it is, in Firefox, the `TypeError` throws as expected at the right time, but the `DOMContentLoaded` callback then fires, long before the shim has a chance to load the module itself:
```
asm.js type error: Asm.js optimizer disabled because no suitable wasm compiler is available [es-module-shims.min-606ae9c3279013fe751cee30f719a592f759e705edb66496812f3d9dbce3d850.js](http://localhost:3000/assets/es-module-shims.min-606ae9c3279013fe751cee30f719a592f759e705edb66496812f3d9dbce3d850.js)
Uncaught TypeError: Error resolving module specifier “munificent_ui/application”. Relative module specifiers must start with “./”, “../” or “/”. [new:27:7](http://localhost:3000/en/fundraisers/1/donations/new)
Uncaught TypeError: window.payments is undefined
    <anonymous> http://localhost:3000/en/fundraisers/1/donations/new:43
    onReady http://localhost:3000/en/fundraisers/1/donations/new:34
    <anonymous> http://localhost:3000/en/fundraisers/1/donations/new:42
[new:43:7](http://localhost:3000/en/fundraisers/1/donations/new)
XHRGEThttp://localhost:3000/assets/munificent_ui/application-3f3cbee6448455fc9bc308b3b87f7ef84ad40996e27701deef188e8c01a31742.js
[HTTP/1.1 200 OK 96ms]
```
```
^^ Module TypeError above is polyfilled and can be ignored ^^ [es-module-shims.js:32:12](http://localhost:3000/dist/es-module-shims.js)
here [application-3f3cbee6448455fc9bc308b3b87f7ef84ad40996e27701deef188e8c01a31742.js:7:9](http://localhost:3000/assets/munificent_ui/application-3f3cbee6448455fc9bc308b3b87f7ef84ad40996e27701deef188e8c01a31742.js)
```

I guess that ideally the polyfill execution would also be timed to happen before `DOMContentLoaded` but I'm not particularly married to waiting on that event if there's an easier way.  I see that `es-module-shims` has a hook that I could use, but presumably I'd not want to use that in e.g. chrome.

Is there a reliable way to wait until after `<script type="module">import "munificent_ui/application"</script>` has loaded, regardless of the polyfill/browser situation?

**Comment by guybedford** [comment]
See https://github.com/guybedford/es-module-shims#no-load-event-retriggers here.

ES Module Shims will retrigger the DOM events for you, but it does mean they execute twice - once when they really fire, and again when ES Module Shims fires them.

As a result, you should find your `onReady` function being called twice in the example when polyfilled. And the second time should give the correct behaviour.

That's the best we can do on this unfortunately, it should definitely be documented better.

**Comment by elliotcm** [comment]
Right I see, OK.  That should be fine, since it does actually work for my code.  I hadn't noticed that it was running the second time.

Thanks!
